src/UglyToad.PdfPig/Outline/BookmarksProvider.cs

```diff
@@ -1,470 +1,442 @@
+                        ExtractNameTree(kidDictionary, catalog, pdfScanner, isLenientParsing, log, explicitDestinations);
+                    }
+                    else if (!isLenientParsing)
+                    {
+                        throw new PdfDocumentFormatException($"Invalid kids entry in PDF name tree: {kid} in {kids}.");
+                    }
+                }
             }
         }
 
-        private static int ParsePageNumber(string goToStr)
+        private static bool TryReadExplicitDestination(IToken value, Catalog catalog, IPdfTokenScanner pdfScanner,
+            ILog log, out ExplicitDestination destination)
         {
-            if (int.TryParse(System.Text.RegularExpressions.Regex.Match(goToStr, "[0-9]+").Value, out int number))
+            destination = null;
+
+            if (DirectObjectFinder.TryGet(value, pdfScanner, out ArrayToken valueArray)
+            && TryGetExplicitDestination(valueArray, catalog, log, out destination))
             {
-                return number + 1;
+                return true;
             }
-            return 0;
+
+            if (DirectObjectFinder.TryGet(value, pdfScanner, out DictionaryToken valueDictionary)
+                && valueDictionary.TryGet(NameToken.D, pdfScanner, out valueArray)
+                && TryGetExplicitDestination(valueArray, catalog, log, out destination))
+            {
+                return true;
+            }
+
+            return false;
         }
 
-        #region Destinations
-        private void GetDestination(ArrayToken destToken, BookmarkNode currentNode)
+        private static bool TryGetExplicitDestination(ArrayToken explicitDestinationArray, Catalog catalog,
+            ILog log,
+            out ExplicitDestination destination)
         {
-            if (destToken == null || destToken.Length == 0)
+            destination = null;
+
+            if (explicitDestinationArray == null || explicitDestinationArray.Length == 0)
             {
-                throw new ArgumentNullException(nameof(destToken), "BookmarksProvider.GetDestination()");
+                return false;
             }
 
-            // 12.3.2.2 Explicit Destinations
-            // Table 151 – Destination syntax
-            var pageToken = destToken[0];
+            int pageNumber;
+
+            var pageToken = explicitDestinationArray[0];
+
             if (pageToken is IndirectReferenceToken pageIndirectReferenceToken)
             {
-                var pageNumber = structure.Catalog.GetPageByReference(pageIndirectReferenceToken.Data).PageNumber;
-                if (pageNumber.HasValue)
-                {
-                    currentNode.PageNumber = pageNumber.Value;
-                }
-                else
-                {
-                    log.Error("BookmarksProvider.GetDestination(): Cannot find page number.");
-                }
+                pageNumber = catalog.GetPageByReference(pageIndirectReferenceToken.Data).PageNumber ?? 1;
```

Comment:
Hi Eliot, 

thanks a lot for your help on that! Looks very good now.
Only comment I have at the moment is that some pdf files throw `NullReferenceException` (see below).

The easy fix is to replace 
`pageNumber = catalog.GetPageByReference(pageIndirectReferenceToken.Data).PageNumber ?? 1;` by 
`pageNumber = catalog.GetPageByReference(pageIndirectReferenceToken.Data)?.PageNumber ?? 1;`

Full error:
`
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=UglyToad.PdfPig
  StackTrace:
   at UglyToad.PdfPig.Outline.BookmarksProvider.TryGetExplicitDestination(ArrayToken explicitDestinationArray, Catalog catalog, ILog log, ExplicitDestination& destination) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\Outline\BookmarksProvider.cs:line 284
   at UglyToad.PdfPig.Outline.BookmarksProvider.TryReadExplicitDestination(IToken value, Catalog catalog, IPdfTokenScanner pdfScanner, ILog log, ExplicitDestination& destination) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\Outline\BookmarksProvider.cs:line 257
   at UglyToad.PdfPig.Outline.BookmarksProvider.ExtractNameTree(DictionaryToken nameTreeNodeDictionary, Catalog catalog, IPdfTokenScanner pdfScanner, Boolean isLenientParsing, ILog log, Dictionary2 explicitDestinations) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\Outline\BookmarksProvider.cs:line 223
   at UglyToad.PdfPig.Outline.BookmarksProvider.ExtractNameTree(DictionaryToken nameTreeNodeDictionary, Catalog catalog, IPdfTokenScanner pdfScanner, Boolean isLenientParsing, ILog log, Dictionary2 explicitDestinations) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\Outline\BookmarksProvider.cs:line 236
   at UglyToad.PdfPig.Outline.BookmarksProvider.ReadNamedDestinations(Catalog catalog, IPdfTokenScanner pdfScanner, Boolean isLenientParsing, ILog log) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\Outline\BookmarksProvider.cs:line 201
   at UglyToad.PdfPig.Outline.BookmarksProvider.GetBookmarks(Catalog catalog) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\Outline\BookmarksProvider.cs:line 47
   at UglyToad.PdfPig.PdfDocument.TryGetBookmarks(Bookmarks& bookmarks) in D:\VS2017\source\repos\BobLd\PdfPig\src\UglyToad.PdfPig\PdfDocument.cs:line 215
`

Reply:
Thanks, I've changed this to skip explicit destinations where the page specifies a reference that doesn't exist in the page tree instead of defaulting to page 1, not sure if this will cause missing bookmarks though, would you be able to test the newest version just to check please?

Reply:
I've tested it and it seems to work properly, but agree with on the missing bookmarks question. 

As an aside, if we ever need to default to a page number, I would chose -1 to signal that this is not an actual page.

Reply:
Thanks for testing that.

That makes sense, the reason I've avoided using the `-1` or `0` approach at the moment is it relies on the consumer knowing that they need to check the page number for some edge condition where it won't be a valid index, for that reason I'd prefer to use `int?` which at least makes the edge condition explicit (though to be fair the -1 approach is fairly common in `IndexOf` etc). But for now I'd rather skip the bookmark entirely, I think there may be a case where the page indirect reference links to the real indirect reference for the page object but I'll merge this for now and squish any bugs as and when they're encountered.